target/linux/ath79/dts/ar7161_buffalo_wzr-hp-ag300h.dts

```diff
@@ -0,0 +1,287 @@
+			gpios = <&ath9k1 3 GPIO_ACTIVE_LOW>;
+			default-state = "off";
+		};
+
+		movie_engine {
+			label = "buffalo:blue:movie_engine";
+			gpios = <&ath9k1 4 GPIO_ACTIVE_LOW>;
+			default-state = "off";
+		};
+
+		band5g_a {
+			label = "buffalo:amber:band5g";
+			gpios = <&ath9k1 5 GPIO_ACTIVE_LOW>;
+			default-state = "off";
+		};
+	};
+
+	gpio-keys-polled {
+		compatible = "gpio-keys-polled";
+		poll-interval = <20>;
+
+		reset {
+			linux,code = <KEY_RESTART>;
+			gpios = <&gpio 11 GPIO_ACTIVE_LOW>;
+			debounce-interval = <60>;
+		};
+
+		usb {
+			linux,code = <BTN_2>;
+			gpios = <&gpio 3 GPIO_ACTIVE_LOW>;
+			debounce-interval = <60>;
+		};
+
+		aoss {
+			linux,code = <KEY_WPS_BUTTON>;
+			gpios = <&gpio 5 GPIO_ACTIVE_LOW>;
+			debounce-interval = <60>;
+		};
+
+		router_auto {
+			linux,code = <BTN_6>;
+			gpios = <&gpio 6 GPIO_ACTIVE_LOW>;
+			debounce-interval = <60>;
+		};
+
+		router_off {
+			linux,code = <BTN_5>;
+			gpios = <&gpio 7 GPIO_ACTIVE_LOW>;
+			debounce-interval = <60>;
+		};
+
+		movie_engine {
+			linux,code = <BTN_7>;
+			gpios = <&gpio 8 GPIO_ACTIVE_LOW>;
+			debounce-interval = <60>;
+		};
+	};
+
+	gpio-export {
+		compatible = "gpio-export";
+
+		gpio_usb_power {
+			gpio-export,name = "buffalo:power:usb";
+			gpio-export,output = <1>;
+			gpios = <&gpio 2 GPIO_ACTIVE_HIGH>;
+		};
+	};
+
+	flash {
+		compatible = "virtual,mtd-concat";
+
+		devices = <&flash0 &flash1>;
+
+		partitions {
+			compatible = "fixed-partitions";
+			#address-cells = <1>;
+			#size-cells = <1>;
+			partition@0 {
+				label = "u-boot";
+				reg = <0x0000000 0x0040000>;
+				read-only;
+			};
+
+			partition@40000 {
+				label = "u-boot-env";
+				reg = <0x0040000 0x0010000>;
+				read-only;
+			};
+
+			art: partition@50000 {
+				label = "art";
+				reg = <0x0050000 0x0010000>;
+				read-only;
+			};
+
+			partition@60000 {
+				label = "firmware";
+				reg = <0x0060000 0x1f90000>;
+			};
+
+			partition@1ff0000 {
+				label = "user_property";
+				reg = <0x1ff0000 0x0010000>;
+				read-only;
+			};
+		};
+	};
+};
+
+&usb_phy {
+	status = "okay";
+};
+
+&usb1 {
+	#address-cells = <1>;
+	#size-cells = <0>;
+	status = "okay";
+
+	usb_ochi_port: port@1 {
+		reg = <1>;
+		#trigger-source-cells = <0>;
+	};
+};
+
+&usb2 {
+	#address-cells = <1>;
+	#size-cells = <0>;
+	status = "okay";
+
+	usb_echi_port: port@1 {
+		reg = <1>;
+		#trigger-source-cells = <0>;
+	};
+};
+
+&pcie0 {
+	status = "okay";
+
+	ath9k0: wifi@0,11 {
+		compatible = "pci168c,0029";
+		reg = <0x8800 0 0 0 0>;
+		qca,no-eeprom;
+		#gpio-cells = <2>;
+		gpio-controller;
+	};
+
+	ath9k1: wifi@0,12 {
+		compatible = "pci168c,0029";
+		reg = <0x9000 0 0 0 0>;
+		qca,no-eeprom;
+		#gpio-cells = <2>;
+		gpio-controller;
+	};
+};
+
+&uart {
+	status = "okay";
+};
+
+&pll {
+	clocks = <&extosc>;
+};
+
+&spi {
+	status = "okay";
+	num-cs = <2>;
+	cs-gpios = <0>, <0>;
+
+	/* Winbond 25Q128BVFG SPI flash */
+	flash0: flash@0 {
+		compatible = "winbond,w25q128", "jedec,spi-nor";
```

Comment:
Use jedec,spi-nor here is enough. The flash used by manufacturer isn't always fixed.

Reply:
Then of course I will get warnings in the logs and can't use the correct frequency and fast-read. 
But if that's the way to go...

Reply:
Albeit [it is valid to add a chip specific compatible ](https://elixir.bootlin.com/linux/v4.14.67/source/Documentation/devicetree/bindings/mtd/jedec,spi-nor.txt) we are usually don't do for the reason mentioned by @981213. Board manufactors tend to change flash chips during the lifetime of a board.

The issue you are describing sound rather like a bug somewhere. The flash should be identified by its JEDEC ID. I don't see any code in [m25p80.c](https://elixir.bootlin.com/linux/v4.14.67/source/drivers/mtd/devices/m25p80.c) or [spi-nor.c](https://elixir.bootlin.com/linux/v4.14.67/source/drivers/mtd/spi-nor/spi-nor.c) which takes the w25q128 compatible into account.

Reply:
I think you're right, the warning probably only happened on ar71xx. I will confirm that later. However, fast-read and frequency are always taken from the dts, not from by the JEDEC id.

How is this handled usually? I see there are some ramips boards that have fast-read set, with varying frequencies, even though they (now) specify compatible = "jedec,spi-nor".
For non-dts setups, spi-nor.c enables fast-read by default, and disables it for some known JEDEC id entries.
Should I enable it and keep the winbond max-frequency? Or disable it and lower frequency to some "default" value?

Reply:
Setting fast read and frequency via devicetree is fine and correct.

Reply:
Changed to only "jedec,spi-nor" in the second version.